### ISSUE TITLE
How to pretrain using Sth-Sth-v2 Dataset

### ISSUE THREAD

**Issue by yangjiangeyjg** [siddk/voltron-robotics#9]
Hi, Siddharth. 
This is a great job and How can I use training code for pretraining?
Thanks!

**Comment by siddk** [comment]
Hey @yangjiangeyjg! I just pushed the full preprocessing pipeline, as well as code for training MVP, R3M, and Voltron models on GCP TPU VMs with PyTorch XLA.

I'm working on a "default" PyTorch training script (on GPUs), but hopefully the logic in `examples/xla-reference/xpretrain.py` is easy enough to follow for now! Let me know if not, and keep an eye on the repository -- I'll push the cleaned pretraining pipeline very soon!

**Comment by zhouliang-yu** [comment]
> 

Hey @siddk 
Have you finished the GPU version?

**Comment by zhouliang-yu** [comment]
> Hey @yangjiangeyjg! I just pushed the full preprocessing pipeline, as well as code for training MVP, R3M, and Voltron models on GCP TPU VMs with PyTorch XLA.
> 
> I'm working on a "default" PyTorch training script (on GPUs), but hopefully the logic in `examples/xla-reference/xpretrain.py` is easy enough to follow for now! Let me know if not, and keep an eye on the repository -- I'll push the cleaned pretraining pipeline very soon!

Hey @siddk, have you finished the GPU version? I find the xpretrain.py still hard to follow, cause I have limited experience on training on TPUs

**Comment by zhouliang-yu** [comment]
> Hey @yangjiangeyjg! I just pushed the full preprocessing pipeline, as well as code for training MVP, R3M, and Voltron models on GCP TPU VMs with PyTorch XLA.
> 
> I'm working on a "default" PyTorch training script (on GPUs), but hopefully the logic in `examples/xla-reference/xpretrain.py` is easy enough to follow for now! Let me know if not, and keep an eye on the repository -- I'll push the cleaned pretraining pipeline very soon!

Hey @siddk, Thx so much for your effort.
Btw, have you finished the GPU version? I find the xpretrain.py still hard to follow, cause I have limited experience in training on TPUs

**Comment by zhouliang-yu** [comment]
> Hey @yangjiangeyjg! I just pushed the full preprocessing pipeline, as well as code for training MVP, R3M, and Voltron models on GCP TPU VMs with PyTorch XLA.
> 
> I'm working on a "default" PyTorch training script (on GPUs), but hopefully the logic in `examples/xla-reference/xpretrain.py` is easy enough to follow for now! Let me know if not, and keep an eye on the repository -- I'll push the cleaned pretraining pipeline very soon!

Hey @siddk, Thx so much for your effort.
Btw, As you mentioned you are working on the GPU version of the training script, I wonder if it is finished, cannot wait to follow the GPU version, since I dun have available access to TPU haha

**Comment by yangjiangeyjg** [comment]
> > Hey @yangjiangeyjg! I just pushed the full preprocessing pipeline, as well as code for training MVP, R3M, and Voltron models on GCP TPU VMs with PyTorch XLA.
> > I'm working on a "default" PyTorch training script (on GPUs), but hopefully the logic in `examples/xla-reference/xpretrain.py` is easy enough to follow for now! Let me know if not, and keep an eye on the repository -- I'll push the cleaned pretraining pipeline very soon!
> 
> Hey @siddk, Thx so much for your effort. Btw, As you mentioned you are working on the GPU version of the training script, I wonder if it is finished, cannot wait to follow the GPU version, since I dun have available access to TPU haha

I have the same problem.

**Comment by yangjiangeyjg** [comment]
> Hey @yangjiangeyjg! I just pushed the full preprocessing pipeline, as well as code for training MVP, R3M, and Voltron models on GCP TPU VMs with PyTorch XLA.
> 
> I'm working on a "default" PyTorch training script (on GPUs), but hopefully the logic in `examples/xla-reference/xpretrain.py` is easy enough to follow for now! Let me know if not, and keep an eye on the repository -- I'll push the cleaned pretraining pipeline very soon!

Thanks for your efforts and I am also looking forward to the GPU version. 
Thanks!

**Comment by aliang8** [comment]
Are you guys able to successfully download the Sth-sth dataset from the Qualcomm website? I downloaded the files but could not untar the compressed files. Any suggestions?

**Comment by siddk** [comment]
Hey @aliang8 - I just tried downloading off of Qualcomm as part of getting instructions/scripts set up for from-scratch pretraining. These are the instructions I had to follow (will push these with full processing/pretraining scripts):

---

Follow the instructions [at this link](https://developer.qualcomm.com/software/ai-datasets/something-something) to download the dataset. Qualcomm requires that you register for a [Qualcomm OneID Account](https://myaccount.qualcomm.com/signup?target=https%3A%2F%2Fdeveloper.qualcomm.com) to get access to the data. Approval might take some time.

After registering for an account, make sure to download all of the following files to a directory of your choosing 
(we create a directory `data/raw/something-something-v2/downloaded/`). *You will need to manually download all 22 of
the following files from the Qualcomm site*:

1. Datasheet / Instructions (PDF – optional, but useful): `20bn-something-something_download_instructions_-_091622.pdf`
2. Labels (includes language annotations): `20bn-something-something_download-package-labels.zip`
3. Chunked Videos (should be 20 `.zip` archives):
   + `20bn-something-something-v2-00.zip`
   + ...
   + `20bn-something-something-v2-19.zip`

To extract all the given files (we extract to `data/raw/something-something-v2/`) - *execute the following from inside
the `downloaded/` subdirectory)*:

```bash
# Labels (annotations/language) --> creates `data/raw/something-something-v2/labels`
unzip 20bn-something-something-download-package-labels.zip -d ../

# Videos (following instructions in `20-bn-something-something_download_instructions_-_091622.pdf`)
unzip "20bn-something-something-v2-*.zip" -d ../videos
cd ../videos
cat 20bn-something-something-?? | tar -xvzf -
find . -maxdepth 1 -type f -delete
cd 20bn-something-something-v2/
find . -mindepth 1 -maxdepth 1 -exec mv -t .. -- {} +
cd ..
rm -r 20bn-something-something-v2
ls | wc   # Should have 220847 `.webm` files!
```

---

Can you try out the above and let me know if it works?

**Comment by aliang8** [comment]
Thank you for the detailed reply! 

I'm running into a problem when executing this line: `cat 20bn-something-something-?? | tar -xvzf -` 

I'm getting this error: 
<img width="440" alt="image" src="https://user-images.githubusercontent.com/16139634/229617664-94c0123c-974d-4347-b317-52cfb43efad3.png">

**Comment by siddk** [comment]
What shell are you using?

**Comment by aliang8** [comment]
I am using zsh and `cat ...` gives me a bunch of unreadable machine text.

**Comment by siddk** [comment]
A single chunk as in just a single file? My understanding of the instructions are that you need all 20 files for this command to work properly.

**Comment by aliang8** [comment]
Oh hmm, okay I will try with downloading all 20 files.

**Comment by aliang8** [comment]
Works! Thank you :)

**Comment by yangjiangeyjg** [comment]
Hi, @siddk, how can I use GPUs training based on pytorch DDP.

**Comment by siddk** [comment]
Hey @yangjiangeyjg @aliang8 @zhouliang-yu; I am so sorry for the delay! Full code for preprocessing/pretraining in native PyTorch (single + multi-GPU) is now on the main branch (via the latest release).

Some useful files:

- General Overview of Preprocessing/Pretraining for Sth-Sth-v2: [examples/pretrain/README.md](https://github.com/siddk/voltron-robotics/blob/main/examples/pretrain/README.md)
- PyTorch DDP Pretraining Script (invoke via `torchrun`): [examples/pretrain/pretrain.py](https://github.com/siddk/voltron-robotics/blob/main/examples/pretrain/pretrain.py)

Sorry again for the delay - please let me know if there are any questions. Extending this to other datasets/models should follow the same flow (following updated preprocessing code at [examples/pretrain/preprocess.py](https://github.com/siddk/voltron-robotics/blob/main/examples/pretrain/preprocess.py).

Closing this issue for now; for different questions, please post another issue/shoot me an email!
